docs/en/creating-widgets/supplemental.md

```diff
@@ -0,0 +1,869 @@
+	private counter: number = 0;
+
+	protected render() {
+		return (
+			<div>
+				Current widget state: {this.myState}
+				<br />
+				<button
+					onclick={() => {
+						this.myState = 'State change iteration #' + ++this.counter;
+					}}
+				>
+					Change State
+				</button>
+			</div>
+		);
+	}
+}
+```
+
+Note that this example is not complete - clicking on the 'Change State' button in the running application will not have any effect on the widget's render output. This is because the state is fully encapsulated within `MyEncapsulatedStateWidget`, and Dojo [is not aware of any changes made to it](#widget-properties). Only the widget's initial render will be processed by the framework.
+
+In order to notify Dojo that a re-render is needed, widgets that encapsulate render state need to invalidate themselves.
+
+### Invalidating a widget
+
+Function-based widgets can use the [`invalidator` middleware](../middleware/supplemental.md#invalidator) to mark themselves as invalid.
+
+For class-based widgets, there are two ways to invalidate:
+
+1.  Explicitly calling `this.invalidate()` in an appropriate location where state is being changed.
+    -   In the `MyEncapsulatedStateWidget` example, this could be done in the 'Change State' button's `onclick` handler.
+2.  Annotating any relevant fields with the `@watch()` decorator (from the `@dojo/framework/core/vdomecorators/watch` module). When `@watch`ed fields are modified, `this.invalidate()` will implicitly be called - this can be useful for state fields that always need to trigger a re-render when updated.
+
+**Note:** marking a widget as invalid won't immediately re-render the widget - instead it acts as a notification to Dojo that the widget has been dirtied, so should be updated and re-rendered in the next render cycle. This means invalidating a widget multiple times within the same render frame won't have any negative impact on application performance, although excessive invalidation should be avoided to ensure optimal performance.
+
+The following is an updated `MyEncapsulatedStateWidget` example that will correctly update its output when its state is changed.
+
+**Function-based variant:**
+
+```tsx
+import { create, tsx, invalidator } from '@dojo/framework/core/vdom';
+
+const factory = create({ invalidator });
+
+let myState: string = 'Hello from a stateful widget!';
+let counter: number = 0;
+
+export default factory(function MyEncapsulatedStateWidget({ middleware: { invalidator } }) {
+	return (
+		<div>
+			Current widget state: {myState}
+			<br />
+			<button
+				onclick={() => {
+					myState = 'State change iteration #' + ++counter;
+					invalidator();
+				}}
+			>
+				Change State
+			</button>
+		</div>
+	);
+});
+```
+
+**Class-based variant:**
+
+Here, both `myState` and `counter` are updated as part of the same application logic operation, so `@watch()` could be added to either or both of the fields, with the same net effect and performance profile in all cases:
```

Comment:
Do you think we should demonstrate using `this.invalidate()` but also talk about the option of using `@watch()`?

Reply:
Using `this.invalidate()` was mentioned above these examples:

> For class-based widgets, there are two ways to invalidate:
> 
> 1.  Explicitly calling `this.invalidate()` in an appropriate location where state is being changed.
>     -   In the `MyEncapsulatedStateWidget` example, this could be done in the 'Change State' button's `onclick` handler.

would it be better to swap usage here? `invalidate` in the example and `@watch` as the alternative in the above description?

Reply:
I think is fine, given the mentions above